### PR TITLE
GKE Versioning Constraint v1

### LINKED ISSUE #268: [FORSETI] Kubernetes Engine Version scanner functionality
Related to Forseti Security Issue [#3624](https://github.com/forseti-security/forseti-security/issues/3624)

### Scanner

Kubernetes Engine Version Scanner

**Description:** Kubernetes Engine clusters running on older versions can be exposed to security vulnerabilities, or lack of support. The KE version scanner can ensure your Kubernetes Engine clusters are running safe and supported versions.

### Python Violations Summary
There are three types (modes) of violations that can be generated:
1) **Check_serverconfig_valid_node_versions: true (default = False)** -  check the node pool versions against the supported version list.

2) **Check_serverconfig_valid_master_versions: true (default = False)** - check the master version against the supported version list.

3) **Allowed_nodepool_versions is not an empty list** - check the node pool versions against the allowed versions list.

### 1:1 Rego | Python Parity Status
There does not seem to be any Rego policies related to GKE versioning. It does not look like the Kubernetes Service Config resource is available on CAI, which is needed to check for supported node pool versions and master versions (#1 and #2 from above). 

Should just write Rego policies that check the master version and node_pool version from a list of user defined supported (allowed) versions instead.

### Rego work to do

Create Rego rule(s) to do the following:

1) Check the master version against the user defined supported (allowed) versions list.

2) Check the node pool versions against the user defined supported (allowed) versions list.

### PR DESCRIPTION
**Pull request by hshin-g**
Resolves #268

This PR creates a GKE Cluster Versioning constraint that checks either the `currentMasterVersion` or `currentNodeVersion` field of a GKE cluster with a list of user defined versions. 

This constraint provides the following features:
- Check `currentMasterVersion` or `currentNodeVersion` against list of user defined versions
- Allowlist/denylist mode
- Exemptions

Also included in this PR:
- Sample constraint
- Tests

### CODE REVIEW

**Comment by hshin-g** [comment]
@morgante @briantkennedy Could this get a review from one of you? Thanks!

File: policies/templates/gcp_gke_cluster_version_v1.yaml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+          properties:
+            mode:
+              type: string
+              enum: [denylist, allowlist]
+              description: "String identifying the operational mode, allowlist or denylist. In allowlist mode,
+              GKE clusters are only allowed to have a master version specified in the 'master_versions' parameter. In denylist mode,
+              resources are allowed any master versions except those listed in the 'master_versions' parameter."
+            version_type:
+              type: string
+              enum: [currentMasterVersion, currentNodeVersion]
```
Comment (morgante) [inline policies/templates/gcp_gke_cluster_version_v1.yaml]: These values seem unnecessarily long/confusing.

```suggestion
              enum: [master, node]
```

File: policies/templates/gcp_gke_cluster_version_v1.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+              description: "String identifying the GKE cluster version type to look for, currentMasterVersion or currentNodeVersion.
+              currentMasterVersion refers to the Kubernetes version for GKE cluster masters. currentNodeVersion refers to the Kubernetes
+              version for GKE cluster nodes."
+            exemptions:
+              type: array
+              items: string
+              description: "Array of GKE clusters to exempt from version restriction. String values in the array
```
Comment (morgante) [inline policies/templates/gcp_gke_cluster_version_v1.yaml]: I'd recommend allowing regex/glob.

File: validator/gke_cluster_version.rego [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	count(matches) == 0
+
+	# Get the version value in the asset
+	target_version_type := params.version_type
+	target_version := get_target_value(asset.resource.data, target_version_type)
+
+	# Check that version is in allowlist/denylist
+	asset_version := params.versions
+	version_matches := {target_version} & cast_set(asset_version)
+	target_version_match_count(params.mode, desired_count)
```
Comment (morgante) [inline validator/gke_cluster_version.rego]:
```suggestion
	desired_count := target_version_match_count(params.mode)
```

File: validator/gke_cluster_version_test.rego [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+import data.test.fixtures.gke_cluster_version.constraints.node_version_allowlist_all as node_version_allowlist_all
+import data.test.fixtures.gke_cluster_version.constraints.node_version_allowlist_none as node_version_allowlist_none
+import data.test.fixtures.gke_cluster_version.constraints.node_version_allowlist_one as node_version_allowlist_one
+import data.test.fixtures.gke_cluster_version.constraints.node_version_allowlist_one_exemption as node_version_allowlist_one_exemption
+import data.test.fixtures.gke_cluster_version.constraints.node_version_denylist_all as node_version_denylist_all
+import data.test.fixtures.gke_cluster_version.constraints.node_version_denylist_none as node_version_denylist_none
+import data.test.fixtures.gke_cluster_version.constraints.node_version_denylist_one as node_version_denylist_one
+import data.test.fixtures.gke_cluster_version.constraints.node_version_denylist_one_exemption as node_version_denylist_one_exemption
+
+# Find all violations on our test cases
```
Comment (morgante) [inline validator/gke_cluster_version_test.rego]: I think this should use the new test utils.

File: policies/templates/gcp_gke_cluster_version_v1.yaml (hunk 1 above)
Comment (hshin-g) [reply policies/templates/gcp_gke_cluster_version_v1.yaml]: These values are the actual field names in CAI and are used to find those properties. Should they still be changed?

File: policies/templates/gcp_gke_cluster_version_v1.yaml (hunk 2 above)
Comment (hshin-g) [reply policies/templates/gcp_gke_cluster_version_v1.yaml]: Ack - PTAL

File: validator/gke_cluster_version.rego (hunk 3 above)
Comment (hshin-g) [reply validator/gke_cluster_version.rego]: Ack - PTAL

File: validator/gke_cluster_version_test.rego (hunk 4 above)
Comment (hshin-g) [reply validator/gke_cluster_version_test.rego]: Ack - PTAL

File: policies/templates/gcp_gke_cluster_version_v1.yaml (hunk 1 above)
Comment (hshin-g) [reply policies/templates/gcp_gke_cluster_version_v1.yaml]: Ok- have changed to "master" and "node" as the values. Let me know your thoughts- thanks!

Comment (morgante) [reply policies/templates/gcp_gke_cluster_version_v1.yaml]: We shouldn't expect users to know the CAI values, so thanks for updating.

File: policies/templates/gcp_gke_cluster_version_v1.yaml [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+              whose asset name has the format '//container.googleapis.com/projects/project-abc/zones/{ANY ZONE}/clusters/{ANY NAME}'.
+              such as:
+              '//container.googleapis.com/projects/project-abc/zones/us-central1-c/clusters/my-cluster'
+              but NOT
+              '//container.googleapis.com/projects/wrong-project/zones/us-central1-c/clusters/wrong-cluster'"
+            versions:
+              type: array
+              items: string
+              description: "Array of versions that GKE clusters are allowed to have,
+              e.g. 1.12.10-gke.17 for currentMasterVersion or 1.4.4 for currentNodeVersion."
```
Comment (morgante) [inline policies/templates/gcp_gke_cluster_version_v1.yaml]: Shouldn't refer to "currentMasterVersion"

```suggestion
              e.g. 1.12.10-gke.17 or 1.4.4
```

**Comment by hshin-g** [comment]
@morgante Thanks for your review! Have added more tests for glob functionality, and cleanup some of the doc descriptions- PTAL and let me know your thoughts.

File: policies/templates/gcp_gke_cluster_version_v1.yaml (hunk 5 above)
Comment (hshin-g) [reply policies/templates/gcp_gke_cluster_version_v1.yaml]: Ack

**Pull request merged**